plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,144 @@
+package com.mapbox.maps.plugin.viewport.experimental
+
+import com.mapbox.maps.plugin.delegates.MapDelegateProvider
+import com.mapbox.maps.plugin.viewport.DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS
+import com.mapbox.maps.plugin.viewport.DEFAULT_STATE_TRANSITION_MAX_DURATION_MS
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.MapboxFollowingState
+import com.mapbox.maps.plugin.viewport.experimental.state.MapboxOverviewState
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportStateChangedObserver
+import com.mapbox.maps.plugin.viewport.experimental.transition.TransitionEndListener
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransitionOptions
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStateChangedObservers = CopyOnWriteArrayList<ViewportStateChangedObserver>()
+
+  override val states = CopyOnWriteArraySet<ViewportState>()
+  override var currentState: ViewportState? = null
+    private set
+
+  override fun addViewportState(state: ViewportState) {
+    states.add(state)
+  }
+
+  override fun removeViewportState(state: ViewportState) {
+    if (currentState == state) {
+      state.cleanUp()
+      currentState = null
+    }
+    states.remove(state)
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      state,
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_STATE_TRANSITION_MAX_DURATION_MS)
+        .apply(stateTransitionOptionsBlock).build(),
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS)
+        .apply(frameTransitionOptionsBlock).build(),
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    states.add(state)
+    currentState?.cleanUp()
+    state.transitionToState(
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+    currentState = state
+    notifyStateChanged()
```

Comment:
Maybe write a custom setter for `currentState` and trigger the notify from there. Would also help with https://github.com/mapbox/mapbox-maps-android/pull/975/files#r768972819

Reply:
cleanup calls could go in that setter too